rekall-core/rekall/plugins/linux/heap_analysis.py

```diff
@@ -0,0 +1,4388 @@
+                self.heap_vmas.append(vma)
+
+        for arena in self.arenas:
+            if arena.is_main_arena:
+                for mmap_chunk in arena.mmapped_first_chunks:
+                    vma = get_vma_for_offset(self.vmas, mmap_chunk.v())
+                    if vma not in self.heap_vmas:
+                        self.heap_vmas.append(vma)
+
+            else:
+                for heap in arena.heaps:
+                    vma = get_vma_for_offset(self.vmas, heap.v())
+                    if vma not in self.heap_vmas:
+                        self.heap_vmas.append(vma)
+
+
+
+    def _check_and_report_non_main_arena(self, chunk, chunk_in_use):
+        """Checks the given chunk for the NON_MAIN_ARENA bit and prints a warning if not set.
+        This functions should obviously only be used with chunks not belonging to main_arena
+        but also not for MMAPPED chunks (they don't have the NON_MAIN_ARENA bit set).
+        """
+
+        if chunk_in_use and not chunk.non_main_arena():
+            self.session.logging.warn(
+                "Unexpected error: The non main arena chunk at offset 0x{0:x} doesn't have the "
+                "NON_MAIN_ARENA bit set.".format(chunk.v()))
+
+
+    def __log_mmapped_warning_messages(self, warning):
+
+        if not self.mp_:
+            self.session.logging.warn(warning)
+
+        else:
+            self.__mmapped_warnings.add(warning)
+
+
+    # As there might be multiple scenarios, in which a vm_area is mistakenly treated
+    # as a mmapped region (see following warn messages for details),
+    # we strictly test for prev_size to be 0x0 (normally always the case for the
+    # first chunk in a memory region), the size to be != 0 and the mmapped bit to be set
+    def _check_and_report_mmapped_chunk(self, mmap_chunk, mmap_vma):
+        """Checks the given chunk for various MMAPPED chunk specific attributes.
+        Depending on the results and the location of the chunk, a info or warning is printed.
+        """
+
+        base_string = "Current MMAPPED chunk at offset 0x{0:x} ".format(mmap_chunk.v())
+
+        zero_first_chunk_error_reasons = (
+            "As this chunk resides at the beginning of the vm_area, "
+            "this fact might have multiple reasons: "
+            "1. It is part of a MMAPPED region but there are not yet any allocated chunks. "
+            "2. The current vm_area is in fact the rest of a dead thread stack or belongs "
+            " to a mapped file, which is not disginguishable from heap-vmas at the moment. "
+            "3. There might be an unexpected error. "
+            "In the first two cases, this warning can be considered harmless.")
+
+        zero_middle_chunk_error_reasons = (
+            "In the current case, this fact might have the following reasons: "
+            "1. It is the result from an MMAPPED region, which doesn't use "
+            "the whole space for its chunks (in this case harmless). "
+            "2. The cur
```

Comment:
Since renderer.write() is deprecated, all output must be in columns. You can use a Divider column to break it up but realize that to Rekall output is generally a table (which may be sorted in any way).

If this part of the code does something different, you can make another plugin.

We do need to think about how to present these results better.